go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -231,6 +232,29 @@ func newSharedGatewayOpenFlowManager(patchPort, macAddress, gwBridge, gwIntf str
 		dftFlows = append(dftFlows,
 			fmt.Sprintf("cookie=%s, priority=50, in_port=%s, ip, "+
 				"actions=ct(zone=%d, table=1)", defaultOpenFlowCookie, ofportPhys, config.Default.ConntrackZone))
+
+		physicalIP, err := util.MatchIPNetFamily(false, ips)
+		if err != nil {
+			return nil, fmt.Errorf("unable to determine IPv4 physical IP of host: %v", err)
+		}
+		// table 0, SVC Hairpin from OVN destined to local host, DNAT and go to table 3
+		dftFlows = append(dftFlows,
+			fmt.Sprintf("cookie=%s, priority=500, in_port=%s, ip, ip_dst=%s, ip_src=%s,"+
+				"actions=ct(commit,zone=%d,nat(dst=%s),table=3)",
+				defaultOpenFlowCookie, ofportPatch, types.V4HostMasqueradeIP, config.Gateway.V4JoinSubnet,
+				secondaryCTZone, physicalIP.IP))
+
+		// table 0, Reply SVC traffic from Host -> OVN, unDNAT and goto table 2
+		dftFlows = append(dftFlows,
+			fmt.Sprintf("cookie=%s, priority=500, in_port=LOCAL, ip, ip_dst=%s,"+
+				"actions=ct(zone=%d,nat,table=2)",
+				defaultOpenFlowCookie, config.Gateway.V4JoinSubnet, secondaryCTZone))
```

Comment:
why ip_dst is not matching V4HostMasqueradeIP given this is not for hairpin traffic?

Reply:
This is for reply traffic that will come from the host back to ovn from a hairpin. So this is the packet lifecycle from beginning:

1. packet from host -> service, goes into OVN, snat to V4HostMasqueradeIP, send to OVN
2. OVN DNAT's to service endpoint (back to the host), and SNATs to V4JoinSubnet, sends packet back to br-ex
3. Flow on line 240 handles this packet and DNATs the packet to the host IP, sends the packet into the host with source IP V4JoinSubnet, dest IP host
4. Host replies to the packet, and sends back a packet destined to V4JoinSubnet, source dest IP host
5. Now this flow gets hit, which checks from LOCAL the dest IP is V4JoinSubnet (now we know this packet must be a reply, so we can unDNAT it, which will change the source back to V4HostMasqueradeIP, and send it back into OVN
6. OVN will have an existing CT session for this, and unSNAT (change dest from v4JoinSubnet back to V4HostMasqueradeIP) and unDNAT (change source back to service IP), then send back to br-ex
7. Packet will hit the flow on line 316, and unSNAT (change dest ip from V4HostMasqueradeIP back to host IP), and send the packet back to the host with src ip == service ip